### DISCUSSION TITLE
Running a Sequence of Animations

### DISCUSSION THREAD

**Discussion by dusevitch** [sofa-framework/sofa#4101] [Write / Run a simulation] [answered]
I've looked at a few previous posts in this area but they haven't seemed to help my application:

- [https://www.sofa-framework.org/community/forum/topic/how-to-define-an-animation-path/](https://www.sofa-framework.org/community/forum/topic/how-to-define-an-animation-path/
)
- [https://www.sofa-framework.org/community/forum/topic/specifying-an-animation/](https://www.sofa-framework.org/community/forum/topic/specifying-an-animation/)

I'm trying to run a sequence of animations where I animate something, then animate something following, not at the same time. Could be similar to following some spline path, but at specific timed increments. An application would be navigating around an obstacle 

I've tried to stack animation calls but both animations run simultaneously. Here's a very simple example:

```python
def animation1(actuators, step, factor):
    rigid = RigidDof(actuators[0].ServoMotor.ServoBody.dofs)
    rigid.setPosition(rigid.rest_position - rigid.up * step * factor)

def animation2(actuators, step, factor):
    rigid = RigidDof(actuators[0].ServoMotor.ServoBody.dofs)
    rigid.setPosition(rigid.rest_position - rigid.left * step * factor)

def animation3(actuators, step, factor):
    rigid = RigidDof(actuators[0].ServoMotor.ServoBody.dofs)
    rigid.setPosition(rigid.rest_position + rigid.up * step * factor)

def runAnimation(): # This would be run in createScene somewhere
    animate(animation1,{"actuators": self.actuators, "step": 70.0},duration=0.2)
    animate(animation2,{"actuators": self.actuators, "step": 70.0},duration=0.2)
    animate(animation2,{"actuators": self.actuators, "step": 70.0},duration=0.2)
```
Also tried to figure out if there was someway to call using the `onDone` function.

What's a better way of doing this that will actually work? I wasn't exactly clear on whether/how the `AnimationManager` or `MultiStepAnimationLoop` played a role in helping this specific case.

**Comment by damienmarchal** [comment, marked as answer]
Hi @dusevitch,

The animate function has an optional parameter "onDone" that can be used to provide a callback function executed when the animation is finished.

Here is the example from animate.py
```py
def createScene(rootNode):
    def myAnimate1(target, factor):
        print("I should do something on: "+target.name+" factor is: "+str(factor))

    def myOnDone(target, factor):
        print("onDone: "+target.name+" factor is: "+str(factor))

    animate(myAnimate1, {"target": rootNode}, 10, onDone=myOnDone)
```

This make it possible to write animation chains with the following pattern: 
```py
def runAnimation(): 
    def startAnimation3(actuators, step, factor):
          animate(animation3,{"actuators":  actuators, "step": step}, duration=0.2)      

    def startAnimation2(actuators, step, factor):
          animate(animation2,{"actuators": actuators, "step": step}, duration=0.2, onDone=startAnimation3)      

    animate(animation1,{"actuators": self.actuators, "step": 70.0},duration=0.2, onDone=startAnimation2)
```
